src/librustc_mir/interpret/snapshot.rs

```diff
@@ -1,31 +1,70 @@
-//! This module contains the machinery necessary to detect infinite loops
-//! during const-evaluation by taking snapshots of the state of the interpreter
-//! at regular intervals.
-
-// This lives in `interpret` because it needs access to all sots of private state.  However,
-// it is not used by the general miri engine, just by CTFE.
-
+//! Capture and compare snapshots of the compile-time interpreter state to detect when a program
+//! will loop infinitely.
+//!
+//! This lives in `interpret` because it needs access to all sots of private state.  However,
+//! it is not used by the general miri engine, just by CTFE.
+//!
+//! # Comparing interpreter state
+//!
+//! If we observe the same interpreter state at two different points in time during const-eval, we
+//! can be certain that the program is in an infinite loop. For `CompileTimeInterpreter`, the
+//! relevant bits of interpreter state are the stack and the heap. We refer to a copy of this state
+//! as a "snapshot".
+//!
+//! While comparing stacks between snapshots is straightforward, comparing heaps between snapshots
+//! is not. This is because `AllocId`s, which determine the target of a pointer, are frequently
+//! thrown away and recreated, even if the actual value in memory did not change. Consider the
+//! following code:
+//!
+//! ```rust,ignore(const-loop)
+//! const fn inf_loop() {
+//!     // Function call to prevent promotion.
```

Comment:
Doh! Can't believe I missed this. My worry is that the test for [issue 52475](https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-eval/issue-52475.rs) isn't really creating new `AllocId`s, since the `&0` is getting promoted out.